File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/WorkflowFacadeSupport.kt
package dev.bota.sdk.internal

import dev.bota.sdk.BotaErrorCode
import dev.bota.sdk.BotaOperation
import dev.bota.sdk.BotaSDKError
import dev.bota.sdk.internal.core.CoreCommand
import dev.bota.sdk.internal.core.CoreNotification
import dev.bota.sdk.internal.core.CoreNotificationKind
import dev.bota.sdk.internal.jni.NativeCoreException
import kotlinx.coroutines.CancellationException

internal suspend fun awaitWorkflowCompletion(command: CoreCommand, runtime: DeviceRuntime) {
    var completed = false
    try {
        runtime.engine.run(command, runtime.capabilities).collect { notification ->
            when (notification.kind) {
                CoreNotificationKind.Completed -> completed = true
                CoreNotificationKind.Cancelled -> throw BotaSDKError.Core(
                    BotaErrorCode.Cancelled,
                    operation(command.kind),
                    retryable = true,
                    protocolStatus = null,
                    detail = "device workflow was cancelled",
                )
                CoreNotificationKind.Failed -> throw notification.workflowError()
                else -> Unit
            }
        }
    } catch (error: BotaSDKError) {
        throw error
    } catch (error: NativeCoreException) {
        throw error.toPublicError()
    }
    if (!completed) throw BotaSDKError.Core(
        BotaErrorCode.Internal,
        operation(command.kind),
        retryable = true,
        protocolStatus = null,
        detail = "device workflow ended without terminal completion",
    )
}

internal suspend fun runCleanupActions(vararg actions: suspend () -> Unit) {
    var firstFailure: Throwable? = null
    actions.forEach { action ->
        try {
            action()
        } catch (failure: Throwable) {
            val first = firstFailure
            if (first == null) firstFailure = failure else first.addSuppressed(failure)
        }
    }
    firstFailure?.let { throw it }
}

internal suspend fun runCleanupAfter(primaryFailure: Throwable?, vararg actions: suspend () -> Unit) {
    try {
        runCleanupActions(*actions)
    } catch (cleanupFailure: Throwable) {
        if (primaryFailure == null) throw cleanupFailure
        primaryFailure.addSuppressed(cleanupFailure)
    }
}

internal fun CoreNotification.requiredText(id: Int, operation: BotaOperation): String =
    packet.texts(id).firstOrNull() ?: throw malformedNotification(id, operation)

internal fun CoreNotification.requiredUnsigned(id: Int, operation: BotaOperation): ULong =
    packet.unsigneds(id).firstOrNull() ?: throw malformedNotification(id, operation)

internal fun CoreNotification.requiredBoolean(id: Int, operation: BotaOperation): Boolean =
    packet.booleans(id).firstOrNull() ?: throw malformedNotification(id, operation)

internal fun Throwable.facadePublicError(operation: BotaOperation): Throwable = when (this) {
    is BotaSDKError -> this
    is CancellationException -> this
    is NativeCoreException -> toPublicError()
    else -> BotaSDKError.Core(
        BotaErrorCode.Internal,
        operation,
        retryable = true,
        protocolStatus = null,
        detail = message ?: "native workflow facade failed",
    )
}

internal fun cancelled(operation: BotaOperation): BotaSDKError.Core = BotaSDKError.Core(
    BotaErrorCode.Cancelled,
    operation,
    retryable = true,
    protocolStatus = null,
    detail = "device workflow was cancelled",
)

private fun malformedNotification(id: Int, operation: BotaOperation): BotaSDKError.Core = BotaSDKError.Core(
    BotaErrorCode.UnexpectedEvent,
    operation,
    retryable = false,
    protocolStatus = null,
    detail = "workflow notification is missing field $id",
)

private fun operation(commandKind: Int): BotaOperation = when (commandKind) {
    0x0104 -> BotaOperation.Provision
    0x0109, 0x010a -> BotaOperation.FactoryReset
    else -> BotaOperation.Unknown(commandKind.toUInt())
}

File: platforms/apple/Sources/BotaAppSDK/WorkflowFacadeSupport.swift
import BotaDeviceSDKC

func awaitWorkflowCompletion(_ command: CoreCommand, runtime: DeviceRuntime) async throws {
    var completed = false
    do {
        let notifications = await runtime.engine.run(command, capabilities: runtime.capabilities)
        for try await notification in notifications {
            switch notification.kind {
            case .completed:
                completed = true
            case .cancelled:
                throw BotaSDKError(
                    code: .cancelled,
                    operation: BotaSDKError.operation(notification.packet.operation),
                    retryable: true,
                    detail: "device workflow was cancelled"
                )
            case .failed:
                throw workflowError(notification)
            case .started, .deviceDiscovered, .connectionEstablished, .progress,
                 .retrying, .deviceUploadPreserved, .bleFallbackReady,
                 .firmwareProgress, .deviceLog, .streamingPaused, .streamingResumed,
                 .streamingCompleted, .encryptedUploadV2Staged:
                break
            }
        }
    } catch let error as BotaSDKError {
        throw error
    } catch let error as CoreError {
        throw BotaSDKError(error)
    }
    guard completed else {
        throw BotaSDKError(
            code: .internal,
            operation: BotaSDKError.operation(command.kind),
            retryable: true,
            detail: "device workflow ended without a terminal completion"
        )
    }
}

func workflowError(_ notification: CoreNotification) -> BotaSDKError {
    let fields = notification.packet.fields
    let rawCode = fields.facadeUnsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_ERROR_CODE)) ?? 21
    return BotaSDKError(
        code: BotaSDKError.code(UInt32(clamping: rawCode)),
        operation: BotaSDKError.operation(notification.packet.operation),
        retryable: fields.facadeBool(UInt32(BOTA_DEVICE_SDK_V1_FIELD_RETRYABLE)) ?? false,
        protocolStatus: fields
            .facadeUnsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_PROTOCOL_STATUS))
            .flatMap { UInt16(exactly: $0) },
        detail: fields.facadeText(UInt32(BOTA_DEVICE_SDK_V1_FIELD_ERROR_DETAIL))
            ?? "device workflow failed"
    )
}

private extension Array where Element == CoreField {
    func facadeUnsigned(_ id: UInt32) -> UInt64? {
        for field in self {
            if case let .unsigned(fieldID, value) = field, fieldID == id { return value }
        }
        return nil
    }

    func facadeBool(_ id: UInt32) -> Bool? {
        for field in self {
            if case let .bool(fieldID, value) = field, fieldID == id { return value }
        }
        return nil
    }

    func facadeText(_ id: UInt32) -> String? {
        for field in self {
            if case let .text(fieldID, value) = field, fieldID == id { return value }
        }
        return nil
    }
}
